src/transport/srv_http.rs

```diff
@@ -112,42 +107,91 @@ impl Transport for SrvHttpTransport {
+                let server_opts = server_opts.clone();
+
                 url.set_path("check-in");
 
-                let span = tracing::trace_span!("check-in attempt", host = url.to_string());
-
-                async move {
-                    tracing::trace!("Fetching check-in configuration.");
-
-                    reqwest
-                        .post(url)
-                        .header(
-                            http::header::CONTENT_TYPE,
-                            crate::transport::APPLICATION_JSON,
-                        )
-                        .body(payload)
-                        .send()
-                        .await
-                        .map_err(SrvHttpTransportError::from)
-                }
-                .instrument(span)
+                let span = tracing::trace_span!("check-in attempt", %url);
+
+                perform_request(reqwest, url, payload, server_opts).instrument(span)
             })
             .await?;
 
-        Ok(resp.json().await?)
+        let checkin: Checkin = resp.json().await?;
+
+        // Update server options to sync up compression options
+        {
+            let mut opts = self.server_options.write().await;
+            *opts = checkin.server_options.clone();
+        }
+
+        Ok(checkin)
     }
 }
 
+#[tracing::instrument(skip(reqwest, payload, server_opts))]
+async fn perform_request(
+    reqwest: reqwest::Client,
+    url: url::Url,
+    payload: Vec<u8>,
+    server_opts: Arc<tokio::sync::RwLock<ServerOptions>>,
+) -> Result<reqwest::Response, SrvHttpTransportError> {
+    let algos = server_opts.read().await.compression_algorithms.into_iter();
+
+    for compression_algo in algos {
+        let span = tracing::debug_span!("requesting", ?compression_algo);
+
+        let mut req = reqwest
+            .post(url.clone())
+            .header(
+                http::header::CONTENT_TYPE,
+                crate::transport::APPLICATION_JSON,
+            )
+            .body(compression_algo.compress(&payload).await?);
+
+        if let Some(encoding) = compression_algo.content_encoding() {
+            req = req.header(http::header::CONTENT_ENCODING, encoding);
+        }
+
+        tracing::trace!(parent: &span, "Requesting");
+        match req.send().instrument(span.clone()).await {
+            Ok(resp) if resp.status() == http::StatusCode::UNSUPPORTED_MEDIA_TYPE => {
+                tracing::debug!(
+                    ?compression_algo,
+                    "Disabling compression algorithm because it is unsupported"
+                );
+                server_opts
+                    .write()
+                    .await
+                    .compression_algorithms
+                    .delete(&compression_algo);
+            }
+
+            Err(e) => {
+                return Err(SrvHttpTransportError::from(e));
+            }
+            Ok(resp) => return Ok(resp),
+        }
+    }
+
+    Err(SrvHttpTransportError::NoCompressionMode)
```

Comment:
I wonder why we make this an error and not just make it the former-default ("identity", I assume)?

Reply:
There are a few cases:

1. The server did not return any information about supported compression modes, and the transport tried `zstd` and then `identity` already,
2. The server explicitly returned a list including `identity` and it was attempted then failed,
3. The server explicitly did _not_ return `identity` in the list, because they don't want clients to try it.